### PR TITLE
Remove inventory overflow

### PR DESCRIPTION
**Pull request by kevingranade**
Fixes #2843
Instead of giving players huge encumbrance penalties because they picked up a rock, simply don't let them pick up the rock.
In support of this, if you drop an article of clothing that takes you under-volume, random items will fall out of your inventory to get you back under the limit again.  same if an item with inventory capacity is destroyed.

### CODE REVIEW

**Comment by Rivet-the-Zombie** [comment]
All of this is far too awesome for words!

**Comment by BevapDin** [comment]
How does that behave when an item changes its volume when it's applied? For example the tanning leather hide changes to a tanned hide, which has a higher volume.

**Comment by kevingranade** [comment]
That will get you overvolume, a simple fix is to call player::drop_inventory_overflow() after any action that might increase your volume, or to be extra paranoid we could just call it after every action.

**Comment by kevingranade** [comment]
I pushed without thinking it through, armor destruction is reactive and
therefore should keep the call.

**Comment by kevingranade** [comment]
Thanks for the feedback BevapDin, I put it at end of turn instead of on every action, partially because I realized it should go in the character class and there's no on action hook for NPCs.  Also I restored the call that makes armor destruction drop items immediately.

**Comment by Coolthulhu** [comment]
The random removal depends on item order. It could be made order independent by making it a roulette, like this:
```
int index_of_selected_item = 0;
int volume_counter = 0;
int random_roll = rng(0, carried_volume);
for( item &i : items ) {
    volume_counter += i.volume();
    if( volume_counter > random_roll ) {
        break;
    }
    index_of_selected_item++;
}
```

While this isn't exactly a major issue, it does change the probability of selecting a given item. For example:
We have 100 items, each of those has 1 volume. The probability of selecting second item is 1%, but then the probability of un-selecting it later is 98%, meaning that the probability of dropping the second item is just 0.02%.

File: src/character.cpp [hunk 1]
```diff
@@ -585,6 +586,16 @@ bool Character::can_pickWeight( int weight, bool safe ) const
     }
 }
 
+void Character::drop_inventory_overflow() {
+    if( volume_carried() > volume_capacity() ) {
+        for( auto &item_to_drop :
+               inv.remove_randomly_by_volume( volume_carried() - volume_capacity() ) ) {
+            g->m.add_item_or_charges( pos(), item_to_drop );
+        }
+        add_msg_if_player( m_bad, _("Some items tumble to the ground.") );
```
Comment (Coolthulhu) [inline src/character.cpp]: This is always "items", even when there's a single item.

**Comment by karlnp** [comment]
Does it still take N amount of turns to drop N items on clothing removal?

**Comment by kevingranade** [comment]
The weighting provided by x_in_y(volume, total_volume) makes selection
random based on volume, unless I made a mistake it will have no bias.
The forced dropping costs no turns.

**Comment by karlnp** [comment]
cool, good. it's been annoying to have to pre-empty spend twenty turns dropping books and food when fighting on the way back from a supply run.

**Comment by Cat09** [comment]
how far is this from being able to pull stuff from your inventory volume to the (containment) volume of an item you just removed? (obviously a small subset of the scenarios where this pull comes into play, nice work Kevin.)

**Comment by kevingranade** [comment]
I don't follow, do you mean choosing which items get dropped along with the
container you just dropped?  If that's what you mean this doesn't get us
any closer, that requires overhauling the inventory to store things in
items instead of in a separate list like it does now.

**Comment by Coolthulhu** [comment]
There could be a menu, like liquid handling one, that would pop up when a container is lost/dropped.

**Comment by Cat09** [comment]
actually, I meant simply dropping anything at random into the container as you drop it
perhaps filling it as close to capacity as possible, or filling by weight and volume, attempting to fill the item with as close to the same ratios as your inventory has

**Comment by kevingranade** [comment]
If you're dropping the container, you should already be able to chose what
gets dropped with it by dropping other items at the same time.  If the
container is destroyed it doesn't make any sense to retroactively choose
what is in it.
What Cat09 is saying is roughly what I was talking about, and DavidKeaton
has been working on it.

**Pull request merged**